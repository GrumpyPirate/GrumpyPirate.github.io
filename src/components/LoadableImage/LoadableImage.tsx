import React, { FunctionComponent, ImgHTMLAttributes, useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';

import { animDur, animEase } from 'styles';
import { ClassNameProps } from 'types/common';

const StyledImg = styled.img<{ isLoaded: boolean }>`
  opacity: 0;
  visibility: hidden;
  transition-property: visibility, opacity;
  transition-duration: ${animDur}s;
  transition-timing-function: ${animEase};

  ${({ isLoaded }) =>
    isLoaded &&
    css`
      opacity: 1;
      visibility: visible;
    `}
`;

/**
 * Renders a Image component.
 *
 * @returns {ReactElement}
 */
const LoadableImage: FunctionComponent<ClassNameProps & ImgHTMLAttributes<{}>> = ({
  className,
  alt = '',
  ...rest
}) => {
  const imgRef = useRef<HTMLImageElement>(null);
  const [isImageLoaded, setImageLoaded] = useState(false);

  useEffect(() => {
    if (imgRef && imgRef.current && imgRef.current.complete) {
      setImageLoaded(true);
    }
  }, []);

  return (
    <StyledImg
      alt={alt}
      className={className}
      data-testid="loadable-image"
      isLoaded={isImageLoaded}
      loading="lazy"
      onLoad={() => {
        setImageLoaded(true);
      }}
      ref={imgRef}
      {...rest}
    />
  );
};

export default LoadableImage;
