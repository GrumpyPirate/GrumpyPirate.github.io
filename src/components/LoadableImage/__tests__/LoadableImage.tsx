import { act, fireEvent, render } from '@testing-library/react';
import React, { ImgHTMLAttributes } from 'react';

import testImageSrc from 'images/portfolio/supporting-images/39mc.jpg';

import LoadableImage from '../LoadableImage';

describe('Components/LoadableImage', () => {
  const props: ImgHTMLAttributes<{}> = {
    src: testImageSrc,
  };

  test('should render as expected, without crashing', () => {
    const { container } = render(<LoadableImage {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should fade the image in, when the load event fires', () => {
    const testProps: ImgHTMLAttributes<{}> = {
      ...props,
      alt: 'Test img element',
    };
    const { container, getByAltText } = render(<LoadableImage {...testProps} />);

    expect(getByAltText('Test img element')).toHaveStyleRule('opacity', '0');
    expect(getByAltText('Test img element')).toHaveStyleRule('visibility', 'hidden');

    act(() => {
      fireEvent.load(getByAltText('Test img element'));
    });

    expect(getByAltText('Test img element')).toHaveStyleRule('opacity', '1');
    expect(getByAltText('Test img element')).toHaveStyleRule('visibility', 'visible');

    expect(container.firstChild).toMatchSnapshot();
  });

  test("should fade the image in, if the image is already cached in the user's browser", () => {
    const spy = jest
      .spyOn(HTMLImageElement.prototype, 'complete', 'get')
      .mockImplementation(() => true);

    const testProps: ImgHTMLAttributes<{}> = {
      ...props,
      src: testImageSrc,
      alt: 'Test img element',
    };
    const { container, getByAltText } = render(<LoadableImage {...testProps} />);

    expect(getByAltText('Test img element')).toHaveStyleRule('opacity', '1');
    expect(getByAltText('Test img element')).toHaveStyleRule('visibility', 'visible');

    expect(container.firstChild).toMatchSnapshot();

    spy.mockRestore();
  });
});
