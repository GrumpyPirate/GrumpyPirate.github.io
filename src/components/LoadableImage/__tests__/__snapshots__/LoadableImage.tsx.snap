// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/LoadableImage should fade the image in, if the image is already cached in the user's browser 1`] = `
.c0 {
  opacity: 0;
  visibility: hidden;
  -webkit-transition-property: visibility,opacity;
  transition-property: visibility,opacity;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  opacity: 1;
  visibility: visible;
}

<img
  alt="Test img element"
  class="c0"
  data-testid="loadable-image"
  loading="lazy"
  src="39mc.jpg"
/>
`;

exports[`Components/LoadableImage should fade the image in, when the load event fires 1`] = `
.c0 {
  opacity: 0;
  visibility: hidden;
  -webkit-transition-property: visibility,opacity;
  transition-property: visibility,opacity;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  opacity: 1;
  visibility: visible;
}

<img
  alt="Test img element"
  class="c0"
  data-testid="loadable-image"
  loading="lazy"
  src="39mc.jpg"
/>
`;

exports[`Components/LoadableImage should render as expected, without crashing 1`] = `
.c0 {
  opacity: 0;
  visibility: hidden;
  -webkit-transition-property: visibility,opacity;
  transition-property: visibility,opacity;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
}

<img
  alt=""
  class="c0"
  data-testid="loadable-image"
  loading="lazy"
  src="39mc.jpg"
/>
`;
